internal/clientconn/conn.go

```diff
@@ -140,7 +146,29 @@ func (c *conn) run(ctx context.Context) (err error) {
 	}()
 
 	bufr := bufio.NewReader(c.netConn)
+
+	// if test record path is set, split netConn reader to write to file and bufr
+	if c.testRecordPath != "" {
+		if err := os.MkdirAll(c.testRecordPath, 0o644); err != nil {
+			return err
+		}
+
+		filename := fmt.Sprintf("%s_%s.log", c.netConn.RemoteAddr().String(), time.Now().Format("2006-02-01_15:04:05"))
+
+		path := filepath.Join(c.testRecordPath, filename)
+
+		//nolint:nosnakecase// os package require to use underscore
+		f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o744)
+		if err != nil {
+			return err
+		}
+
```

Comment:
⚠️ We should close that file; otherwise, some data may be lost.

That looks like a right place to do it there with `defer f.Close()`, but please check it

Reply:
Nice catch, definitely something that shouldn't be forgotten. I checked and it looks like everything works fine with defer in this place.